src/goto-instrument/contracts/dynamic-frames/dfcc_check_loop_normal_form.cpp

```diff
@@ -0,0 +1,203 @@
+/*******************************************************************\
+
+Module: Dynamic frame condition checking for loop contracts
+
+Author: Qinheping Hu, qinhh@amazon.com
+Author: Remi Delmas, delmasrd@amazon.com
+
+\*******************************************************************/
+
+#include "dfcc_check_loop_normal_form.h"
+
+#include <analyses/natural_loops.h>
+#include <goto-instrument/contracts/utils.h>
+
+void dfcc_check_loop_normal_form(goto_programt &goto_program, messaget &log)
+{
+  natural_loops_mutablet natural_loops(goto_program);
+
+  // used to keep track of loop heads across different loops
+  std::set<goto_programt::targett> head_set;
+
+  // instruction span for each loop
+  std::vector<std::pair<goto_programt::targett, goto_programt::targett>> spans;
+
+  for(auto &loop : natural_loops.loop_map)
+  {
+    auto head = loop.first;
+    // Check that no two loops share the same head
+    if(head_set.find(head) != head_set.end())
+    {
+      log.error() << "At least two loops share the same head: "
+                  << head->source_location() << messaget::eom;
+      throw analysis_exceptiont("Found loops with the same head");
+    }
+    head_set.insert(head);
+
+    auto &loop_instructions = loop.second;
+
+    if(loop_instructions.size() <= 1)
+    {
+      // Ignore single instruction loops of the form
+      // `LOOP: IF cond GOTO LOOP;`
+      continue;
+    }
+
+    auto latch = head;
+    bool latch_found = false;
+
+    // Find latch and check it is unique
+    for(const auto &t : loop_instructions)
+    {
+      if(t->is_goto() && t->get_target() == head)
+      {
+        if(latch_found)
+        {
+          // clang-format off
+          log.error() << "Loop starting at:"
+                      << "\n- head: " << head->source_location()
+                      << "\nhas more than one latch instruction:"
+                      << "\n- latch1: " << latch->source_location()
+                      << "\n- latch2: " << t->source_location()
+                      << messaget::eom;
+          // clang-format on
+          throw analysis_exceptiont(
```

Comment:
A curious question. Are the two checks "one loop with two latch instructions" and "two loops sharing the same head instruction" equivalent?

Reply:
Yes I think they ought to be, will simplify